spec/src/main/asciidoc/providers.asciidoc

```diff
@@ -47,6 +47,28 @@ The `ReaderInterceptor` interface is a listener for when a read occurs against t
 
 The `WriterInterceptor` interface is a listener for when a write occurs to the stream to be sent on the remote service invocation.
 
+=== ResponseExceptionMapper
+
+The `ResponseExceptionMapper` is specific to MicroProfile Rest Client.  This mapper will take a `Response` object retrieved via an invocation of a client and convert it to a `Throwable`, if applicable.  The runtime should scan all of the registered mappers, sort them ascending based on `getPriority()`, find the ones that can handle the given `Response`, and invoke them.  The first one discovered where `toThrowable` returns a non-empty `Optional` object will be thrown by the runtime.
+
+==== How to Implement ResponseExceptionMapper
+
+The specification provides default methods for `getPriority()` and `handles(Response)` methods.  Priority is meant to be derived via a `@Priority` annotation added to the client.  The runtime will sort ascending, taking the one with the lowest numeric value first to check if it can handle the `Response` object.  The usage of ascending sorting is done to be consistent with JAX-RS behavior.
+
+Likewise, the `handles` method by default will handle any response code >= 400.  You may override this behavior if you so choose to handle other response codes (both a smaller ranger and a larger range are expected).
+
+The `toThrowable(Response)` method actually does the conversion work.  This method should not raise any `Throwable`, instead just return a `Throwable` if it can.  It returns `Optional<Throwable>` in case it just wants to do a payload inspection.
+
+Any methods that read the response body as a stream must ensure that they reset the stream.
+
+==== Built in ResponseExceptionMapper
+
+Each implementation will provide out of the box a `ResponseExceptionMapper` implementation that will map the response into a `WebApplicationException` whenever the response status code is >= 400.  It has a priority of `Integer.MAX_VALUE`.  It is meant to be used as a fall back whenever an error is encountered.  This mapper will be registered by default to all client interfaces.
+
```

Comment:
Just a thought.... would it make sense to disable the built-in mapper if the user has registered their own?

Here is the scenario - the user considers only specific HTTP responses to be an exception (i.e. 401, 403, 500) but others are ok (i.e. 404 might just indicate that a key is missing from a hash map or something similar - not necessarily an exception).  They create and register one or more of their own custom mappers, but then they would also need to explicitly disable the built-in mapper to avoid getting a WebApplicationException when a non-exception HTTP response code is returned.

What do you think?  Other options include:
1) Don't automatically register the built-in mapper - but make it available in the spec APIs so that users could easily add it.
2) Make some kind of annotation-based switch to disable the built-in mapper, something like `@ReplacesBuiltIn` so that a user doesn't need to add any config.

Anyway, just something to think about - this should not hold up delivery.

Reply:
You mean instead of the property?  The reasoning behind it is to make it simple to setup from an API consumer's point of view.  I'm assuming that the consumer is defining the interface, not the producer.  I'd rather see it used by default without a user having to do anything else.

Reply:
My thought was more along the lines of "if you (user) do not register a ResponseExceptionMapper, then the framework will provide a default one (>=400 --> WebApplicationException)", but "if you do register any mapper at all, then the framework will not provide the default one".

I'm starting to doubt whether my idea is actually all that intuitive though... I think it would mean less code/config for some cases, but might be more confusing for some customers.